### PR TITLE
TST: Tests and Helpers for Datetime/Period Arrays

### PR DESCRIPTION
**Pull request by jbrockmendel**
Implement ABCDatetimeArray, ABCTimedeltaArray

Fix some problems in DatetimeArray comparison methods, with tests

Implement assert_datetime_array_equal, still need to do the same for TimedeltaArray

Extend fixtures to allow us to run some of the arithmetic tests on PeriodArray, DatetimeArray.  Change a few usages to use these fixtures.

This is a companion-PR to #23493.  Without that one, we can't easily extend the existing tests.  Without this one, problems in the comparison methods would cause failures.

### CODE REVIEW

File: pandas/tests/arithmetic/conftest.py [hunk 1]
```diff
@@ -171,3 +172,21 @@ def box_df_broadcast_failure(request):
     the DataFrame operation tries to broadcast incorrectly.
     """
     return request.param
+
+
+@pytest.fixture(params=[pd.Index, pd.Series, pd.DataFrame, PeriodArray],
+                ids=lambda x: x.__name__)
+def pbox(request):
```
Comment (jreback) [inline pandas/tests/arithmetic/conftest.py]: can u make a bit longer names

File: pandas/tests/arrays/test_datetimes.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        dti = pd.date_range('2016-01-1', freq='MS', periods=9, tz=None)
+        arr = DatetimeArray(dti)
+        assert arr.freq == dti.freq
+        assert arr.tz == dti.tz
+
+        right = dti
+
+        expected = np.ones(len(arr), dtype=bool)
+
+        for op in [operator.eq, operator.le, operator.ge]:
```
Comment (jreback) [inline pandas/tests/arrays/test_datetimes.py]: can u parametrize

File: pandas/util/testing.py [hunk 3]
```diff
@@ -1049,6 +1051,18 @@ def assert_period_array_equal(left, right, obj='PeriodArray'):
     assert_attr_equal('freq', left, right, obj=obj)
 
 
+def assert_datetime_array_equal(left, right, obj='DatetimeArray'):
+    _check_isinstance(left, right, DatetimeArray)
```
Comment (jreback) [inline pandas/util/testing.py]: why is this needed? why don’t we just have a more generic assert_array_equal

adding more comparison routines generally is just asking for trouble - we already have too many

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/arrays/test_datetimes.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimes.py]: sure

File: pandas/util/testing.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/util/testing.py]: I think we may be able to de-duplicate these once DTA/TDA are EAs, but for now the options are to have this function or to implement something equivalent _inside_ `tm.assert_equal`

File: pandas/tests/arithmetic/conftest.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/arithmetic/conftest.py]: sure

File: pandas/core/arrays/datetimes.py [hunk 4]
```diff
@@ -118,27 +118,36 @@ def wrapper(self, other):
         else:
             if isinstance(other, list):
                 # FIXME: This can break for object-dtype with mixed types
-                other = type(self)(other)
-            elif not isinstance(other, (np.ndarray, ABCIndexClass, ABCSeries)):
+                try:
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimes.py]: Comment in previous line should be removed, will update

File: pandas/core/arrays/datetimes.py [hunk 5]
```diff
@@ -118,27 +118,36 @@ def wrapper(self, other):
...
                 # e.g. is_timedelta64_dtype(other)
                 return ops.invalid_comparison(self, other, op)
             else:
                 self._assert_tzawareness_compat(other)
-                result = meth(self, np.asarray(other))
+                if not hasattr(other, 'asi8'):
+                    # ndarray, Series
+                    other = type(self)(other)
+                result = meth(self, other)
+                o_mask = other._isnan
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/datetimes.py]: `isna(other)` ?

File: pandas/tests/arithmetic/conftest.py [hunk 6]
```diff
@@ -171,3 +172,21 @@ def box_df_broadcast_failure(request):
...
+def box_with_period(request):
+    """
+    Like `box`, but specific to PeriodDtype for also testing PeriodArray
+    """
+    return request.param
+
+
+@pytest.fixture(params=[pd.Index, pd.Series, pd.DataFrame, DatetimeArray],
+                ids=lambda x: x.__name__)
+def box_with_dtarray(request):
```
Comment (jorisvandenbossche) [inline pandas/tests/arithmetic/conftest.py]: `box_with_datetime` ? (to be consistent with the period name)

File: pandas/util/testing.py [hunk 7]
```diff
@@ -1049,6 +1051,18 @@ def assert_period_array_equal(left, right, obj='PeriodArray'):
...
+
+    assert_numpy_array_equal(left._data, right._data,
+                             obj='{obj}._data'.format(obj=obj))
+    assert_attr_equal('freq', left, right, obj=obj)
+    assert_attr_equal('tz', left, right, obj=obj)
+
+    # Check that == works as expected
+    assert ((left == right) | (left._isnan & right._isnan)).all()
```
Comment (jorisvandenbossche) [inline pandas/util/testing.py]: why is this needed? (we shouldn't be testing the equality operator inside the assert? or otherwise simply use this as the initial assert if it is sufficient?)

File: pandas/util/testing.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/util/testing.py]: I think it is fine to reconsider this once all internal ExtensionArrays are in place (I think the same was said when adding the `assert_period_array_equal` that is just above this one)

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good to me in general, some small comments

**Comment by jorisvandenbossche** [comment]
> This is a companion-PR to #23493. Without that one, we can't easily extend the existing tests. Without this one, problems in the comparison methods would cause failures.

But do they depend on each other? (meaning, does one need to be merged before the other, or they are orthogonal implementation-wise?)

File: pandas/tests/arithmetic/conftest.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/tests/arithmetic/conftest.py]: sure

File: pandas/core/arrays/datetimes.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: It turns out that `isna(DatetimeArray)` doesn't work at the moment.  I want to keep the footprint of these PRs contained as long as possible, so chose to avoid changing it for now.

File: pandas/util/testing.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/util/testing.py]: I included this as a double-check because ATM the comparison operations on DatetimeArray are broken.  When we have a robust set of tests for those comparisons, this will be entirely unnecesary.  I don't mind removing it if it bothers you, but would personally prefer to leave it until those tests are in place.

**Comment by jbrockmendel** [comment]
> But do they depend on each other? (meaning, does one need to be merged before the other, or they are orthogonal implementation-wise?)

They are orthogonal implementation-wise.  They are both needed in order to make major progress on sharing existing tests.

File: pandas/core/dtypes/generic.py [hunk 8]
```diff
@@ -53,6 +53,10 @@ def _check(cls, inst):
                                         ('sparse_array', 'sparse_series'))
 ABCCategorical = create_pandas_abc_type("ABCCategorical", "_typ",
                                         ("categorical"))
+ABCDatetimeArray = create_pandas_abc_type("ABCDatetimeArray", "_typ",
+                                          ("datetimearray"))
+ABCTimedeltaArray = create_pandas_abc_type("ABCTimedeltaArray", "_typ",
+                                           ("timedeltaarray"))
```
Comment (jbrockmendel) [inline pandas/core/dtypes/generic.py]: needs tests in tests.dtypes.test_generic

File: pandas/util/testing.py (hunk 7 above)
Comment (jorisvandenbossche) [reply pandas/util/testing.py]: What do you mean with "comparison operations are broken" ?

File: pandas/tests/arithmetic/conftest.py (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/conftest.py]: Can you do this one?

File: pandas/core/dtypes/generic.py (hunk 8 above)
Comment (jorisvandenbossche) [reply pandas/core/dtypes/generic.py]: And did you add such tests?

File: pandas/core/arrays/datetimes.py [hunk 9]
```diff
@@ -166,6 +176,9 @@ class DatetimeArrayMixin(dtl.DatetimeLikeArrayMixin):
     # by returning NotImplemented
     timetuple = None
 
+    # ensure that operations with numpy arrays defer to our implementation
+    __array_priority__ = 1000
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/datetimes.py]: Are there tests in here that need this?

File: pandas/core/arrays/datetimes.py (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimes.py]: But you are adding the `isna` method in this PR (to DatetimelikeArrayMixin above), shouldn't it then work?

File: pandas/util/testing.py (hunk 3 above)
Comment (TomAugspurger) [reply pandas/util/testing.py]: > adding more comparison routines generally is just asking for trouble - we already have too many
 
I really like having separate ones, so that you can assert that you aren't accidentally comparing two of the wrong kind of object. If you just have a generic `assert_array_equal` you could have built two ndarrays on accident and there's no way to check that the type is correct (unless you pass that as a parameter?)

Comment (jreback) [reply pandas/util/testing.py]: again not sure why this is; the point is we are comparing versus an expected value
so this just extra verbose

eg we recently consolidatednto assert_equal to avoid this exact prome
now adding these back is just more code 

sure it’s slightly more explicit but IMHO is not worth the extra functions and maintenance over time

Comment (TomAugspurger) [reply pandas/util/testing.py]: > comparing versus an expected value

sometimes though, that expected value isn't constructed explicitly. Sometimes it's the result of some computation (think of the tests in ops for example).

> eg we recently consolidatednto assert_equal to avoid this exact prome
now adding these back is just more code

`assert_equal` just dispatches to each of the specialized asserts. The code has to live somewhere :) But, I don't mean to nitpick over this. I would just rather type

```python
tm.assert_period_array_equal(a, b)
```

than

```python
tm.assert_array_equal(a, b, kind='period')
```

or, worse, `tm.assert_array_equal(a, b)`, because I'm lazy, and accidentally not have a period array.

Comment (jreback) [reply pandas/util/testing.py]: it’s fine

just want to reduce maintenance burden

File: pandas/core/arrays/datetimes.py (hunk 5 above)
Comment (TomAugspurger) [reply pandas/core/arrays/datetimes.py]: `isna(other)` may not work until `DatetimelikeArray(mixin)` inherits from ExtensionArray. I suspect there's a check in `isna` that only dispatches for EAs.

File: pandas/core/arrays/datetimes.py (hunk 9 above)
Comment (TomAugspurger) [reply pandas/core/arrays/datetimes.py]: Do we eventually want this on DatetimelikeArray (the most base class) so that all the arrays inherit the same value? It looks like PeriodArray doesn't define it or `__array_ufunc__` yet (there's an open issue IIRC).

I don't see any tests between `op(ndarray, datetimearray)`, but I may be missing them.

File: pandas/core/arrays/datetimes.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: > isna(other) may not work until `DatetimelikeArray(mixin)` inherits from ExtensionArray. I suspect there's a check in `isna` that only dispatches for EAs.

Generally right.  The check also catches a couple other things.

File: pandas/tests/arithmetic/conftest.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/tests/arithmetic/conftest.py]: Will do this immediately, good catch

File: pandas/util/testing.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/util/testing.py]: > What do you mean with "comparison operations are broken" ?

on master:

```
dti = pd.date_range('2016-01-01', periods=3)
arr = DatetimeArrayMixin(dti)

>>> arr == arr
array([False, False, False])

>>> arr < dti
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/arrays/datetimes.py", line 125, in wrapper
    return ops.invalid_comparison(self, other, op)
  File "pandas/core/ops.py", line 913, in invalid_comparison
    .format(dtype=left.dtype, typ=type(right).__name__))
TypeError: Invalid comparison between dtype=datetime64[ns] and DatetimeArrayMixin
```

This, others in the arithmetic ops, and I'm assuming others I haven't found yet, are why I am so adamant that we get the tests working on DTA/TDA as stand-alone objects before worrying about the composition and/or constructor names.

File: pandas/core/dtypes/generic.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/generic.py]: Not yet, this a note to myself about something I missed.

**Comment by TomAugspurger** [comment]
There seemed to be an install error with one of the windows workers. Triggered a rebuild.

File: pandas/util/testing.py (hunk 7 above)
Comment (jorisvandenbossche) [reply pandas/util/testing.py]: I see.
Still, I would simply remove the check here already now. We will shortly make sure the comparison operations work, and will add proper tests then.

(still I would

**Comment by jbrockmendel** [comment]
rebuild green

File: pandas/util/testing.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/util/testing.py]: done

**Comment by jbrockmendel** [comment]
Travis failure looks wonky and unrelated
